cmd/zfs/zfs_main.c

```diff
@@ -1488,6 +1497,35 @@ zfs_do_destroy(int argc, char **argv)
 			cb.cb_recurse = B_TRUE;
 			cb.cb_doclones = B_TRUE;
 			break;
+		case 't':
+			types = 0;
+			types_specified = B_TRUE;
+			while (*optarg != '\0') {
+				static char *type_subopts[] = { "snapshot",
+				    "snap", "bookmark", NULL };
+
+				switch (getsubopt(&optarg, type_subopts,
+				    &value)) {
+				case 1:
+				case 2:
+					types |= ZFS_TYPE_SNAPSHOT;
+					break;
+				case 3:
+					types |= ZFS_TYPE_BOOKMARK;
+					break;
+				default:
+					(void) fprintf(stderr,
+					    gettext("invalid type '%s'\n"),
+					    value);
+					usage(B_FALSE);
+				}
+			}
+			break;
```

Comment:
A version of this which handles all of the types exists in `zfs_do_get` and `zfs_do_list`.  If you're comfortable with a slightly larger change, it'd be nice if they could all use the same helper function to set `types`.